### DISCUSSION TITLE
APIClient 설계 방향 제안

### DISCUSSION THREAD

**Discussion by seongmin221** [GEON-PPANG/GEON-PPANG-iOS#258] [Ideas]
현재 저희 서버와의 통신 API 구조는 Moya 에서 제공하는 가이드라인을 따르고 있습니닷
`TargetType` 설정 - `MoyaProvider` 관련 설정 및 - `MoyaProvider` 사용해서 적절한 request 전송

하지만 리팩토링 과정에서 Combine + Repository Pattern 을 적용할 생각이고,
이제는 빠른 개발을 위해 사용한 Moya보다 URLSession이 더 나을 것 같다는 생각이 들어
URLSession으로 리팩토링 하기로 결정 되었습니다

<br>

저희가 Moya를 쓰면서는 실제 통신을 보내는 `APIClient`라 할 것 없이 냅다 `MoyaProvider`를 사용하면 됐지만,
URLSession으로 넘어가기 위해 APIClient를 설계해야 합니다

이번 Discussion은 이런 배경에서 APIClient의 설계를 어떻게 하는게 좋을지에 대한 제안사항입니다

하단 이미지는 저엉말 대강 그려둔 설계 제안이구요, 자세한 사항은 추후에 정리해서 다시 올려두겠습니다

<img width="891" alt="image" src="https://github.com/GEON-PPANG/GEON-PPANG-iOS/assets/72431640/c868500d-4a40-4723-a595-63f83edd35ee">

간략하게 설명을 적어두겠습니닷
- `Request`와 그에 대한 `Response` 쌍을 이루어 관계를 나타내고 싶었음
- `Endpoint`라는 것을 클라-서버와 맞닿아 있는 곳으로 생각해서 HTTPMethod, Header, Body, ... 등등 통신에 필요한 요소들을 묶어두고 싶음
- 이렇게 정의된 Endpoint에 적절한 Request와 Response를 매칭
- 결과적으로는 `Request-Endpoint-Response` 형태의 관계가 될듯

**Comment by seongmin221** [comment]
위처럼 설계를 해 보았으나 ... `Endpoint`라는 것의 뜻을 잘 못 알고 있었기에 ㅠ 약간 수정이 필요할 듯 합니다

수정이 필요하다고 생각이 든 이유와 그 생각의 흐름
- `Request`라는 것은 우리가 `struct` 로 구현해 둔 DTO가 아니다
- `Request` 자체는 요청을 의미하고, 요청에 필요한 요소들은 HTTPMethod, Header, Body 등등 이전에 `Endpoint`에 들어갔던 것들이다
- 따라서 이전 설계에서 `Request`가 `Endpoint`를 흡수(?)하고, `Response`와의 관계를 나타내는건 어떨까?

이 내용은 하단 링크와 거의 유사한 구조입니닷
https://medium.com/makingtuenti/writing-a-scalable-api-client-in-swift-4-b3c6f7f3f3fb

**Reply by jeongdung-eo** [reply]
오오 프로토콜 지향 API 클라이언트를 만들어서 기존의 moya에서의 문제점?이욧던 열거형 보완하자는거 맞나여 ?

**Reply by seongmin221** [reply]
움 사실 POP 라고 할 수 있을지는 잘 모르겠다만 ... 
( POP 를 잘 모르기도 하고 단순히 Protocol을 사용하는 것만으로 POP 라고 하기에는 어려울 것 같아서 )

기존에 Moya를 사용한 저희 방법은 Request와 Response 사이에 관계를 나타내지 못했다는 점이라고 생각합니다 !
그래서 이번 방법은 오히려 객체지향이랑 더 가깝다고 생각이 드는게, Request와 Response 사이의 관계를 더 잘 나타낼 수 있기 때문이죠 !

프로토콜을 사용하는 이유는 Interface 로서의 역할(의존성 역전)을 할 수 있기 때문이라고 생각합니다 !

POP를 좀 더 깊게 공부해보고 싶긴 해서 ... 

<img src="https://github.com/GEON-PPANG/GEON-PPANG-iOS/assets/72431640/1cc0fe77-2d06-450b-8b13-a9ddbf9a0b5d" width="400">

이런 책을 사긴 했다만 아직 읽진 않은 ... ㅋ
